webknossos/webknossos/dataset/storage.py

```diff
@@ -0,0 +1,360 @@
+import re
+import warnings
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from os.path import relpath
+from pathlib import Path
+from typing import Any, Dict, Iterator, Optional, Type
+
+import numpy as np
+import wkw
+import zarr
+from numcodecs import Blosc
+
+from webknossos.geometry import BoundingBox, Vec3Int, Vec3IntLike
+
+
+def is_power_of_two(num: int) -> bool:
+    return num & (num - 1) == 0
+
+
+class StorageArrayException(Exception):
+    pass
+
+
+@dataclass
+class StorageArrayInfo:
+    data_format: str
+    num_channels: int
+    voxel_type: np.dtype
+    chunk_size: Vec3Int
+    chunks_per_shard: Vec3Int
+    compression_mode: bool = False
+
+    @property
+    def shard_size(self) -> Vec3Int:
+        return self.chunk_size * self.chunks_per_shard
+
+
+class StorageArray(ABC):
+    data_format = ""
+
+    _path: Path
+
+    def __init__(self, path: Path):
+        self._path = path
+
+    @property
+    @abstractmethod
+    def info(self) -> StorageArrayInfo:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def try_open(_cls, path: Path) -> Optional["StorageArray"]:
+        for cls in (WKWStorageArray, ZarrStorageArray):
+            array_maybe = cls.try_open(path)
+            if array_maybe is not None:
+                return array_maybe
+        return None
+
+    @classmethod
+    @abstractmethod
+    def create(cls, path: Path, storage_info: StorageArrayInfo) -> "StorageArray":
+        pass
+
+    @abstractmethod
+    def read(self, offset: Vec3IntLike, shape: Vec3IntLike) -> np.ndarray:
+        pass
+
+    @abstractmethod
+    def write(self, offset: Vec3IntLike, data: np.ndarray) -> None:
+        pass
+
+    @abstractmethod
+    def resize(
+        self, new_shape: Vec3IntLike, align_with_shards: bool = True, warn: bool = False
+    ) -> None:
+        pass
+
+    @abstractmethod
+    def list_bounding_boxes(self) -> Iterator[BoundingBox]:
+        pass
+
+    @abstractmethod
+    def close(self) -> None:
+        pass
+
+    @staticmethod
+    def get_class(data_format: str) -> Type["StorageArray"]:
+        for cls in (WKWStorageArray, ZarrStorageArray):
+            if cls.data_format == data_format:
+                return cls
+        raise ValueError(f"Data format `{data_format}` is invalid.")
+
+
+class WKWStorageArray(StorageArray):
+    data_format = "wkw"
+
+    _cached_wkw_dataset: Optional[wkw.Dataset]
+
+    def __init__(self, path: Path):
+        super().__init__(path)
+        self._cached_wkw_dataset = None
+
+    @classmethod
+    def try_open(cls, path: Path) -> Optional["WKWStorageArray"]:
+        if (path / "header.wkw").is_file():
+            return cls(path)
+        return None
+
+    @property
+    def info(self) -> StorageArrayInfo:
```

Comment:
Not sure if this should be cached? Same goes for https://github.com/scalableminds/webknossos-libs/pull/627/files#diff-3fe81ffa3877537cf9b035880c7881ae77f6cf7ec59fd1a59c68206d71eeaddcR254